Add lock component documentation (#1756)
OttoWinter/esphomedocs: components/lock/template.rst

Template Lock
===============

.. seo::
    :description: Instructions for setting up template locks that can execute arbitrary actions when locked, unlocked, or opened
    :image: description.svg

The ``template`` lock platform allows you to create simple locks out of just actions and
an optional value lambda. Once defined, it will automatically appear in Home Assistant
as a lock and can be controlled through the frontend.

.. code-block:: yaml

    # Example configuration entry
    lock:
      - platform: template
        name: "Template Lock"
        lambda: |-
          if (id(some_binary_sensor).state) {
            return true;
          } else {
            return false;

>           }

Reviewer: This should probably use the enums instead of true/false

  This will make the Home Assistant frontend show buttons for both LOCK and UNLOCK actions, instead
  of hiding one of them when the lock is LOCKED/UNLOCKED. Defaults to ``false``.
- **id** (*Optional*, :ref:`config-id`): Manually specify the ID used for code generation.
- All other options from :ref:`Lock <config-lock>`.

.. _lock-template-publish_action:

``lock.template.publish`` Action
----------------------------------

You can also publish a state to a template lock from elsewhere in your YAML file
with the ``lock.template.publish`` action.

.. code-block:: yaml

    # Example configuration entry
    lock:
      - platform: template
        name: "Template Lock"
        id: template_lock1

    # in some trigger
    on_...:
      - lock.template.publish:
          id: template_lock1
          state: ON

      # Templated
      - lock.template.publish:
          id: template_lock1
          state: !lambda 'return true;'

Configuration options:

- **id** (**Required**, :ref:`config-id`): The ID of the template lock.
- **state** (**Required**, boolean, :ref:`templatable <config-templatable>`):

>   The state to publish.

Reviewer: State here should match the enums used in config validation.

.. note::

>     This action can also be written in lambdas, the parameter of the `public_state` method denotes if
>     the lock is currently locked or unlocked:

Reviewer: ```suggestion
    This action can also be written in lambdas, the parameter of the `publish_state` method denotes the state the lock should become:
```

    .. code-block:: cpp

>         id(template_lock1).publish_state(false);

Reviewer: ```suggestion
        id(template_lock1).publish_state(lock::LOCK_STATE_LOCKED);
```
